tests/testthat/helper-expect-metadata.R

```diff
@@ -0,0 +1,246 @@
+        validate_effects(
+          filter_types,
+          plot_setting$options$effect,
+          paste0("plot: '",  plot,
+                 "', setting: '", plot_setting$id,
+                 "', option: '", option$id, "'"),
+          default_filter_map)
+      }
+    }
+  }
+  invisible(TRUE)
+}
+
+#' Validate effects
+#'
+#' Effects have filterId and stateFilterId
+#' stateFilterId is the unique ID for the dropdown
+#' filterId is the unique ID of the FilterType, used to identify options
+#' this is because there can be multiple dropdowns using the same underlying
+#' data e.g. colour and size indicator in the bubble plot
+#' setFilter effects have the mapping between filterId and stateFilterId
+#' all other effects refer only to the stateFilterId
+#'
+#' @param filter_types Named filter types
+#' @param effects Effects to check
+#' @param context Additional context appended to any test failure messages
+#' @param state_filter_map Map of state filter ID to filter ID. See above for
+#'   details. Used to check that effects refer to a filter which exists.
+#'
+#' @return Nothing, called for side effect
+#' @noRd
+validate_effects <- function(filter_types, effects, context,
+                             state_filter_map = list()) {
+  if (!is.null(effects$setFilters)) {
+    filter_ids <- vcapply(effects$setFilters, "[[", "filterId")
+    expect_filter_exists(names(filter_types),
+                         filter_ids,
+                         paste(context, "setFilters"))
+    state_filter_ids <- vcapply(effects$setFilters, "[[", "stateFilterId")
+    state_filter_map <- c(state_filter_map,
+                          setNames(filter_ids, state_filter_ids))
+    expect_state_filter_id_unique(state_filter_map)
+  } else if (length(state_filter_map) == 0) {
+    ## This is kind of gross but logic is, if setFilters is not included
+    ## in an effect, then use all the available filters. So have to replicate
+    ## this here for validation
+    state_filter_map <- setNames(names(filter_types), names(filter_types))
+  }
+  if (!is.null(effects$setMultiple)) {
+    expect_filter_exists(names(filter_types),
+                         state_filter_map[unlist(effects$setMultiple)],
+                         paste(context, "setMultiple"))
+  }
+  if (!is.null(effects$setFilterValues)) {
+    expect_filter_exists(names(filter_types),
+                         state_filter_map[names(effects$setFilterValues)],
+                         paste(context, "setFilterValues"))
+    for (state_filter_id in names(effects$setFilterValues)) {
+      filter_id <- state_filter_map[[state_filter_id]]
+      if (isTRUE(filter_types[[filter_id]])) {
+        break
+      }
+      all_option_ids <- get_filter_option_ids(filter_types[[filter_id]]$options)
+      expect_option_exists(all_option_ids,
+                           effects$setFilterValues[[state_filter_id]],
+                           paste(context, "setFilterValues"),
+                           effect$id)
```

Comment:
where is `effect$id` coming from?

Reply:
agh this is some old junk from previous loop which I missed. Should be `state_filter_id` now, thanks.